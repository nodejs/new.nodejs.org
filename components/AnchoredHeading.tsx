import type { FC, PropsWithChildren } from 'react';

type HeadingLevel = 1 | 2 | 3 | 4 | 5 | 6;

type AnchorHeadingProps = PropsWithChildren<{
  level: HeadingLevel;
  id?: string;
}>;

/**
 * This module will automatically replace the header style words
 * (such as `#`,`##`...ect) to an anchor name.
 *
 * This processor will make all the English characters (26 letters)
 * into their lower case, for others in the middle of words,
 * they will become `-` in default. For more than one `-` together,
 * there's only one left.
 *
 * If your title has some non-English characters, please use
 * `<!---->` to quote your English anchor name inside, with your
 * own title beside it. Otherwises it will return you what it is.
 */

// React will ignore "<!" and ">" in the string,
// so we can just use '-- --' to quote the anchor name inside it.
const COMMENT_FOR_HEADANCHOR = /--\x20?([\w\x20-]+)\x20?--/;

const AnchoredHeading: FC<AnchorHeadingProps> = ({ level, id, children }) => {
  const HeadingLevelTag = `h${level}` as any;

  let sanitizedId =
    id ?? children?.toLocaleString().toLocaleLowerCase().replace(/\x20/g, '-');

  if (sanitizedId) {
    const foundAnchorAndTitle = COMMENT_FOR_HEADANCHOR.exec(sanitizedId);

    if (foundAnchorAndTitle) {
      sanitizedId = foundAnchorAndTitle[1]
        .replace(/(\[([^\]]+)]\([^)]+\))/g, '$2')
        .replace(/[^\w]+/g, '-')
        .replace(/[\x20]+/g, '-')
        .replace(/-{2,}/g, '-')
        .replace(/(^-|-$)/g, '');
    }
  }

  return (
    <HeadingLevelTag id={sanitizedId}>
      {children}
      <a
        id={`header-${sanitizedId}`}
        className="anchor"
        href={`#${sanitizedId}`}
        aria-labelledby={sanitizedId}
      />
    </HeadingLevelTag>
  );
};

export default AnchoredHeading;
