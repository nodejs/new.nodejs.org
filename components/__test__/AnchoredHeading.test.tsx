import AnchoredHeading from '../AnchoredHeading';
import { render } from '@testing-library/react';
import React from 'react';

describe('AnchoredHeading', () => {
  it('should render correctly level 1', () => {
    const { container } = render(
      <AnchoredHeading level={1}>Hello</AnchoredHeading>
    );
    expect(container).toMatchSnapshot();
  });

  it('should render correctly level 2', () => {
    const { container } = render(
      <AnchoredHeading level={2}>Hello</AnchoredHeading>
    );
    expect(container).toMatchSnapshot();
  });

  it('should render correctly level 3', () => {
    const { container } = render(
      <AnchoredHeading level={3}>Hello</AnchoredHeading>
    );
    expect(container).toMatchSnapshot();
  });

  it('should render correctly level 4', () => {
    const { container } = render(
      <AnchoredHeading level={4}>Hello</AnchoredHeading>
    );
    expect(container).toMatchSnapshot();
  });

  it('should render correctly level 5', () => {
    const { container } = render(
      <AnchoredHeading level={5}>Hello</AnchoredHeading>
    );
    expect(container).toMatchSnapshot();
  });

  it('should render correctly level 6', () => {
    const { container } = render(
      <AnchoredHeading level={6}>Hello</AnchoredHeading>
    );
    expect(container).toMatchSnapshot();
  });
});
