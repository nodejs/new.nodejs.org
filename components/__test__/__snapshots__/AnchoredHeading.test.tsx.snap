// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnchoredHeading should render correctly level 1 1`] = `
<div>
  <h1
    id="hello"
  >
    Hello
    <a
      aria-labelledby="hello"
      class="anchor"
      href="#hello"
      id="header-hello"
    />
  </h1>
</div>
`;

exports[`AnchoredHeading should render correctly level 2 1`] = `
<div>
  <h2
    id="hello"
  >
    Hello
    <a
      aria-labelledby="hello"
      class="anchor"
      href="#hello"
      id="header-hello"
    />
  </h2>
</div>
`;

exports[`AnchoredHeading should render correctly level 3 1`] = `
<div>
  <h3
    id="hello"
  >
    Hello
    <a
      aria-labelledby="hello"
      class="anchor"
      href="#hello"
      id="header-hello"
    />
  </h3>
</div>
`;

exports[`AnchoredHeading should render correctly level 4 1`] = `
<div>
  <h4
    id="hello"
  >
    Hello
    <a
      aria-labelledby="hello"
      class="anchor"
      href="#hello"
      id="header-hello"
    />
  </h4>
</div>
`;

exports[`AnchoredHeading should render correctly level 5 1`] = `
<div>
  <h5
    id="hello"
  >
    Hello
    <a
      aria-labelledby="hello"
      class="anchor"
      href="#hello"
      id="header-hello"
    />
  </h5>
</div>
`;

exports[`AnchoredHeading should render correctly level 6 1`] = `
<div>
  <h6
    id="hello"
  >
    Hello
    <a
      aria-labelledby="hello"
      class="anchor"
      href="#hello"
      id="header-hello"
    />
  </h6>
</div>
`;
